### DISCUSSION TITLE
Support for DMN (Decision Model Notation)

### DISCUSSION THREAD

**Discussion by todurnatesh** [gorules/zen#115] [Ideas]
Hi
   I would like to know if there can be a built in translator/ support to DMN ( https://www.omg.org/dmn/ ). This is an industry standard and there are a few decision engines which use DMN. It would be nice to have this feature or maybe its already in the future roadmap.

**Comment by ivanmiletic** [comment]
Hi @todurnatesh 

Thanks for the question. We have a lot of experience working with DMN in the past. GoRules was built to solve issues with DMN and some of its engines. If you think about it, we use similar concepts (Graphs and Tables).

These are the pain points that we solve:

1. Horizontal Scalability - it would be quite difficult to maintain DMN engines such as Drools in High Availability mode due to start time and the fact that (as far as we know), Drools needs to load the decisions during the startup in memory. Meaning it is not built as a stateless system. This is good and performant if you have few decisions, but as you scale to 100s or 1000s of different models the startup time of drools will be heavily impacted.  GoRules ZEN Engine is built for Real-Time parsing/compiling and evaluating models in stateless mode - and it does it in sub millisecond for the simple models.
2. Non-Linear Decision - GoRules is already beyond DMN's capabilities in terms of nonlinear decision models having a Switch/Branching node built-in.
3. Programming Languages - Main DMN engines are built in Java (having versions built in dotNet, Node, and others). To maintain it, all engines will need to be coded separately. We have built single core in Rust and will expose it to other languages via native layers. At the moment we support NodeJS and Python, dotNet is coming at the end of Q2 and Java end of the year.
4. Size of the models - DMN uses XML while we use JSON - deserializing is much faster with JSON.
5. Slow Innovation - Even though DMN is backed by companies such IBM, RedHat, and Camunda etc, the evolution of standard is moving extremely slow, for example expanding decision graphs, adding new node types etc. Besides recently added Switch Node, we will be adding a Custom Node that will allow you to invoke your custom code.

Having said that, we might take a look if we can support converting DMN XML models to GoRules JDM - it might not be simple as we have some different concepts.

Best regards,
Ivan

**Reply by OldFarmer86** [reply]
@ivanmiletic Regarding your second point: DMN engines are typically used together with BPMN, so the argument about non-linear decisions doesn’t really hold. BPMN already provides solid branching capabilities.
